### PR TITLE
bpf: Add extension for running sock LB on MKE-related containers

### PR DESCRIPTION
**Pull request by borkmann**
See commit msg.

### CODE REVIEW

**Comment by borkmann** [comment]
test-me-please

**Comment by borkmann** [comment]
test-me-please

**Comment by borkmann** [comment]
test-me-please

**Review by joamaki — APPROVED** [review]
Looks reasonable given the constraints in this situation.

**Comment by borkmann** [comment]
test-me-please

**Comment by borkmann** [comment]
(rebased)

**Review by michi-covalent — APPROVED** [review]

**Comment by borkmann** [comment]
(extended commit message to describe other approaches that have been evaluated as well just for the record)

**Comment by borkmann** [comment]
test-me-please

**Review by joestringer — APPROVED** [review]
Thanks for the well-written explanation in the commit message. Seems like most reasonable option we have for supporting MKE today, and this code is pretty self-contained. I can't wait for some combination of socket cookies or cgroup v2 to let us solve this problem without platform-specific tricks like this.

File: daemon/cmd/kube_proxy_replacement.go [hunk 1]
```diff
@@ -346,6 +350,22 @@ func initKubeProxyReplacementOptions() (bool, error) {
...
+			if h := probesManager.GetHelpers("cgroup_sock_addr"); h != nil {
+				if _, ok := h["bpf_get_cgroup_classid"]; ok {
+					foundClassid = true
+				}
+				if _, ok := h["bpf_get_netns_cookie"]; ok {
+					foundCookie = true
+				}
+			}
+			if !foundClassid || !foundCookie {
+				log.Fatalf("BPF kube-proxy replacement under MKE with --%s needs kernel 5.7 or newer", option.EnableMKE)
```
Comment (brb) [inline daemon/cmd/kube_proxy_replacement.go]: Shouldn't we panic only when `strict == true`?

File: daemon/cmd/kube_proxy_replacement.go [hunk 2]
```diff
@@ -597,6 +621,52 @@ func disableNodePort() {
 	option.Config.EnableHostLegacyRouting = true
 }
 
+func markHostExtension() {
```
Comment (brb) [inline daemon/cmd/kube_proxy_replacement.go:636]: Small comment why this function is needed for those who won't read the commit msg (also, when e.g., after refactoring `git blame` won't point to the original commit) is useful.

File: daemon/cmd/kube_proxy_replacement.go [hunk 3]
```diff
@@ -346,6 +350,22 @@ func initKubeProxyReplacementOptions() (bool, error) {
 			}
 		}
 
+		if option.Config.EnableMKE {
```
Comment (brb) [inline daemon/cmd/kube_proxy_replacement.go]: Shouldn't this be in the `if option.Config.EnableHostReachableServices {` branch above??

File: daemon/cmd/kube_proxy_replacement.go [hunk 4]
```diff
@@ -597,6 +621,52 @@ func disableNodePort() {
 	option.Config.EnableHostLegacyRouting = true
 }
 
+func markHostExtension() {
+	prefix := option.Config.CgroupPathMKE
+	if prefix == "" {
+		mountInfos, err := mountinfo.GetMountInfo()
+		if err != nil {
+			log.WithError(err).Fatal("Cannot retrieve mount infos")
```
Comment (brb) [inline daemon/cmd/kube_proxy_replacement.go]: s/Cannot retrieve mount infos/Cannot retrieve mount infos for MKE/

File: daemon/cmd/kube_proxy_replacement.go [hunk 5]
```diff
@@ -597,6 +621,52 @@ func disableNodePort() {
...
+				if prefix != "" && prefix != mountInfo.MountPoint {
+					log.Fatalf("Multiple cgroup v1 net_cls mounts: %s, %s",
+						prefix, mountInfo.MountPoint)
+				}
+				prefix = mountInfo.MountPoint
+			}
+		}
+	}
+	if prefix == "" {
+		log.Fatal("Cannot retrieve cgroup v1 net_cls mount info")
```
Comment (brb) [inline daemon/cmd/kube_proxy_replacement.go]: Ditto.

File: daemon/cmd/kube_proxy_replacement.go [hunk 6]
```diff
@@ -597,6 +621,52 @@ func disableNodePort() {
...
+	log.Infof("Using cgroup v1 net_cls mount on MKE: %s", prefix)
+	err := filepath.Walk(prefix,
+		func(path string, info os.FileInfo, err error) error {
+			if err != nil {
+				return err
+			}
+			if !info.IsDir() || strings.Contains(path, "kubepods") || path == prefix {
+				return nil
+			}
+			log.Infof("Marking %s as MKE host extension", path)
```
Comment (brb) [inline daemon/cmd/kube_proxy_replacement.go]: Please use the structured logging i.e., `WithField(logfields.Path: path).Info("Marking MKE host extension"`.

File: daemon/cmd/kube_proxy_replacement.go [hunk 7]
```diff
@@ -597,6 +621,52 @@ func disableNodePort() {
...
+			f, err := os.OpenFile(path+"/net_cls.classid", os.O_RDWR, 0644)
```
Comment (brb) [inline daemon/cmd/kube_proxy_replacement.go:669]: Shall we warn if net_cls.classid is non zero?

**Review by brb — CHANGES_REQUESTED** [review]
The logic LGTM, a few nits.

Also, maybe it makes sense to extend the commit msg in order to mention the Istio usecase wrt the socket-lb bypass (that's how we discovered the issue) in the following paragraph:

> This is different compared to regular K8s deployments where such components
reside in the hostns. For the socket LB which is enabled in KPR=strict deployments
this is problematic as these containers are not seen the same way as hostns and
therefore not all the translations might take place as we perform in hostns (like
accessing NodePort via loopback/local addressing, etc).

**Comment by borkmann** [comment]
test-me-please

**Comment by borkmann** [comment]
test-me-please

**Review by brb — APPROVED** [review]

**Comment by borkmann** [comment]
retest-gke

**Pull request merged**